src/main/scala/org/apache/spark/mllib/linalg/distributed/BetterBlockMatrix.scala

```diff
@@ -0,0 +1,188 @@
+    private def rightBlock(i: Int, j: Int, context: TaskContext): Matrix =
+      block(r, rPartitioner, context, i, j)
+
+    def compute(split: Partition, context: TaskContext): Iterator[((Int, Int), Matrix)] = {
+      val row = split.index % rowBlocks
+      val col = split.index / rowBlocks
+      val rowsInThisBlock: Int = (if (row + 1 == rowBlocks && rowsRemainder != 0) rowsRemainder else rowsPerBlock)
+      val colsInThisBlock: Int = (if (col + 1 == colBlocks && colsRemainder != 0) colsRemainder else colsPerBlock)
+
+      val finalResult = DenseMatrix.zeros(rowsInThisBlock, colsInThisBlock)
+      var i = 0
+      while (i < nProducts) {
+        val leftMat = leftBlock(row, i, context)
+        val rightMat = rightBlock(i, col, context)
+        (leftMat, rightMat) match {
+          case (x: DenseMatrix, y: DenseMatrix) =>
+            multiplyAccumulateWithExtension(x, y, finalResult)
+          case _ =>
+             throw new SparkException(s"No support for multiplying: ${leftMat.getClass} by ${rightMat.getClass}.")
+        }
+
+        i += 1
+      }
+
+      Iterator.single(((row, col), finalResult))
+    }
+
+    protected def getPartitions: Array[Partition] =
+      partitions
+
+    /** Optionally overridden by subclasses to specify how they are partitioned. */
+    @transient override val partitioner: Option[Partitioner] =
+      Some(new GridPartitioner(rowBlocks, colBlocks, 1, 1))
+  }
+
+  case class IntPartition(index: Int) extends Partition { }
+
+  private def gridPartition(m: BlockMatrix): (BlockMatrix, GridPartitioner) = {
+    val p = GridPartitioner(m.numRowBlocks, m.numColBlocks)
+    (new BlockMatrix(m.blocks.partitionBy(p), m.rowsPerBlock, m.colsPerBlock, m.numRows(), m.numCols()),
+      p)
+  }
+
+  private def ensureGridPartitioning(m: BlockMatrix): (BlockMatrix, GridPartitioner) = l.blocks.partitioner match {
+    case Some(gp: GridPartitioner) if (gp.rowsPerPart == 1 && gp.colsPerPart == 1) =>
+      (l, gp)
+    case Some(gp: GridPartitioner) =>
+      logDebug(s"Repartitioning a matrix (slow), $l, with a grid partitioner that didn't have 1 block per partition, had: ${gp.rowsPerPart} x ${gp.colsPerPart}")
+      gridPartition(l)
+    case Some(p) =>
+      logDebug(s"Repartitioning a matrix (slow), $l, with a non-grid partitioner: $p")
+      gridPartition(l)
+    case None =>
+      logDebug(s"Partitioning a matrix (slow), $l, with a no partitioner")
+      gridPartition(l)
+  }
+
+  def multiply(preL: BlockMatrix, preR: BlockMatrix): BlockMatrix = {
+    require(preL.numCols() == preR.numRows(),
+      s"""The number of columns of theleft matrix and the number of rows of the right
+          matrix must be equal: ${l.numRows()} x ${l.numCols()}, ${r.numRows()}
+          x ${r.numCols()}. If you think they should be equal, try setting the
+          dimensions of A and B explicitly while initializing them.""".stripMargin)
+    // FIXME: don't require same blocksize on each matrix
```

Comment:
I'm actually OK with this.  I think we'll settle on a reasonable (1K or 4K or whatever) block size which amortizes the overhead of getting to and making calls to the vectorized matrix operations and which also satisfy the memory requirements, and then we'll use it everywhere.